docs/trex_oauth.md

```diff
@@ -0,0 +1,182 @@
+---
+title: Add OAuth to Dashboard Extensions
+layout: docs
+---
+
+If you want your dashboard extension to use OAuth for authentication, you need to be aware of some important changes that were introduced along with sandboxed extensions in Tableau 2019.4. All dashboard extensions are now wrapped in an `iframe`. Because most OAuth sign-in pages enforce the `X-Frame-Options` settings, which helps prevent [clickjacking](https://tools.ietf.org/html/draft-ietf-oauth-v2-23#section-10.13) attacks, you can't load an OAuth sign-in page within the dashboard, or open a dashboard extension dialog window to use for OAuth sign in.
+
+As a best practice, dashboard extensions should open an external browser window when you need to request end-user authorization. In Tableau Server and Tableau Online, you can do this by using methods like `window.open()` and `window.postMessage()`. However, the `window` methods alone will not work in Tableau Desktop, as the external browser window you open for OAuth sign in has no direct connection to the instance of the dashboard extension. This adds a layer of security and requires extra steps to properly route the access token back to your dashboard extension. The following section describes one possible way of coordinating the OAuth interaction using web sockets for communication and the authorization code grant. For security, the client secret should be stored on the server and not in the client code. This is the only safe way to ensure that someone can't use the client credentials to impersonate your application. To simplify your code, your extension should use the same OAuth flow for Tableau Server, Tableau Online, and Tableau Desktop.
```

Comment:
I don't believe we ever distinguish Tableau Server and Tableau Online in other places do we? Can we just say Tableau Server and Tableau Desktop in this paragraph?

Reply:
We do in some places. Mostly in the Tableau Server and Tableau Online Help docs because of the server and and site settings. Here it doesn't make much of a difference, so I'll drop it to save some words.